### PR TITLE
Combined writing for FSTree

### PR DESCRIPTION
**Pull request by roman-khimov**
We can think of replacing peapod with it.

### CODE REVIEW

**Comment by roman-khimov** [comment]
Notice that this scheme also helps with #1951 and #2107 (!) if we're to store meta separately.

File: pkg/local_object_storage/blobstor/fstree/fstree.go [hunk 1]
```diff
@@ -346,6 +379,38 @@ func (t *FSTree) GetBytes(addr oid.Address) ([]byte, error) {
...
+		lengthOff = addrOff + addrSize
+		dataOff   = lengthOff + lengthSize
+	)
+
+	var notFound bool
+
+	for len(data) > dataOff && data[0] == combinedPrefix {
+		notFound = true // The file _is_ combined, so the object _must_ be there.
+		var l = binary.BigEndian.Uint32(data[lengthOff:dataOff])
+		if bytes.Compare(data[addrOff:lengthOff], addr[:]) == 0 {
```
Comment (carpawell) [inline pkg/local_object_storage/blobstor/fstree/fstree.go:375]: why not `bytes.Equal`?

File: pkg/local_object_storage/blobstor/fstree/fstree.go [hunk 2]
```diff
@@ -346,6 +379,38 @@ func (t *FSTree) GetBytes(addr oid.Address) ([]byte, error) {
 	return dec, nil
 }
 
+func extractCombinedObject(addr oid.ID, data []byte) ([]byte, error) {
```
Comment (carpawell) [inline pkg/local_object_storage/blobstor/fstree/fstree.go:359]: `addr` is not `oid.ID`, would not name it this way

File: pkg/local_object_storage/blobstor/fstree/fstree.go (hunk 1 above)
Comment (carpawell) [inline pkg/local_object_storage/blobstor/fstree/fstree.go:375]: it was not supposed to get bytes this way... we do not follow our own rules... it is either the rules are bad, or we should use `addr.Encode` here

File: pkg/local_object_storage/blobstor/fstree/fstree_write_linux.go [hunk 3]
```diff
@@ -29,19 +58,159 @@ func newSpecificWriteData(root string, perm fs.FileMode, noSync bool) func(strin
...
+
+func (b *syncBatch) wait() error {
+	<-b.ready
+	return b.err
+}
+
+func (b *syncBatch) write(id oid.ID, p string, data []byte) error {
+	var (
+		err  error
+		pref [1 + len(id) + 4]byte
```
Comment (carpawell) [inline pkg/local_object_storage/blobstor/fstree/fstree_write_linux.go:126]: we have consts for this numbers but they are defined in a func for some reason

**Review by carpawell — COMMENTED** [review]
No changelog at all; the last commit was not DCOed.

So what is the peapod's fate?

**Comment by roman-khimov** [comment]
> So what is the peapod's fate?

TBD

File: pkg/local_object_storage/blobstor/fstree/fstree.go (hunk 2 above)
Comment (roman-khimov) [reply pkg/local_object_storage/blobstor/fstree/fstree.go:359]: This naming is already used in some places and OID is an address in some ways.

File: pkg/local_object_storage/blobstor/fstree/fstree.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/local_object_storage/blobstor/fstree/fstree.go:375]: Current code is obviously more efficient, even more importantly it's symmetric to `(*syncBatch).write`, so unless the type changes radically we don't care much about the contents.

File: pkg/local_object_storage/blobstor/fstree/fstree_write_linux.go (hunk 3 above)
Comment (roman-khimov) [reply pkg/local_object_storage/blobstor/fstree/fstree_write_linux.go:126]: It's harder to handle in the reader, more code with more fancy calculations. I had a thought about moving them into upper-layer definitions, but they'd need some `combined` prefix that would make them _so_ long. But can be done.

File: pkg/local_object_storage/blobstor/fstree/fstree.go (hunk 2 above)
Comment (roman-khimov) [reply pkg/local_object_storage/blobstor/fstree/fstree.go:359]: Fixed though.

File: pkg/local_object_storage/blobstor/fstree/fstree.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/local_object_storage/blobstor/fstree/fstree.go]: Fixed

**Comment by carpawell** [comment]
>> So what is the peapod's fate?
>
> TBD

I am asking cause we have to rework some things if there is no "small" storage anymore.

Have you considered GET bench? Isn't it slower to read a combined object for only a part of it if there are a few parallel GET requests? Like reading 8MB for 100 bytes only 128 times _can be slower_ than reading 100 bytes 128 times from a peadod.

**Comment by roman-khimov** [comment]
> Isn't it slower to read

For sure. But I doubt it's noticeable. I'll try to bench it.

**Comment by roman-khimov** [comment]
Ironically, it doesn't have any conflicts with current master even without a rebase. So my proposal is to merge this as is and then deal with peapod. @cthulhu-rider, @carpawell?

**Review by cthulhu-rider — APPROVED** [review]

**Comment by carpawell** [comment]
> Have you considered GET bench? Isn't it slower to read a combined object for only a part of it if there are a few parallel GET requests? Like reading 8MB for a 4KB object 128 times can be slower than reading 4KB 128 times from a peadod.

@roman-khimov, my questions are the same. I am afraid there (at least this PR) is not enough info about if it is really a _better_ one than we had before, from _any_ perspective (i consider it a worse solution for GETs). If there is a single slower thing in results, we should discuss it and find out (at least write it somewhere) what is better for us.

**Comment by roman-khimov** [comment]
But you've seen the test results. GETs are the same, PUTs are ~15% better for the target cases. GETs do incur some additional overhead, but it's totally acceptable, there is no real difference in practice.

**Comment by roman-khimov** [comment]
Rebased, otherwise lens failed to build after the merge (even though the merge itself is clean).

**Review by carpawell — APPROVED** [review]

**Pull request merged**